File: MainSource/C++/CompareableLine.cpp
#include "CompareableLine.h"

CompareableLine::CompareableLine() {}
CompareableLine::~CompareableLine() {}

CompareableLine::CompareableLine(Vec4i vPoint) {
	this->vPoint = vPoint;
	iLine_size = (int)pow(vPoint[0] - vPoint[2], 2) + (int)pow(vPoint[1] - vPoint[3], 2);
	if (vPoint[2] == vPoint[0]) dSlope = 90;
	dSlope = atan(-1 * (double)(vPoint[3] - vPoint[1]) / (vPoint[2] - vPoint[0])) * 180.0 / CV_PI;
	if (dSlope < 0) dSlope += 180;
}

Vec4i CompareableLine::getPoint() {
	return vPoint;
}

int CompareableLine::getLine_size() {
	return iLine_size;
}

double CompareableLine::getFunctionS(CompareableLine* ll) {
	if (getFunctionD(ll->getSlope()) == 0) return -10000000;
	return cdParameter_beta*(pow(this->iLine_size + ll->getLine_size(),2) - sdAvgCubeLineSize - getFunctionD(ll->getSlope())) + getFunctionD(ll->getSlope()); // represent loss function of line which involves probablity of loss.
}

double CompareableLine::getFunctionD(double dSlope) {
	if (10 + this->dSlope < dSlope || this->dSlope - 10 > dSlope) return abs(this->dSlope + dSlope - 2 * sdMapDegree)*(this->dSlope + dSlope);
	return 0;
}

bool CompareableLine::calParams() {
	return true;
}

double CompareableLine::getSlope() {
	return dSlope;
}

void CompareableLine::setAvgLineSize(double AvgLineSize) {
	sdAvgLineSize = AvgLineSize;
}

double CompareableLine::getAvgLineSize() {
	return sdAvgLineSize;
}

void CompareableLine::setAvgCubeLineSize(double AvgCubeLineSize) {
	sdAvgCubeLineSize = AvgCubeLineSize;
}

CompareableLine CompareableLine::operator=(const CompareableLine& p) {
	return p;
}

void CompareableLine::setNowDegree(double dNowDegree) {
	sdMapDegree = dNowDegree;
}

void CompareableLine::setParam(double val)
{
	CompareableLine::cdParameter_beta = val;
}

double CompareableLine::sdAvgLineSize;
double CompareableLine::sdAvgCubeLineSize;
double CompareableLine::sdMapDegree;
double CompareableLine::cdParameter_beta;

File: Android/MainSource/Java/CompareableLine.java

class CompareableLine{
  //상수
  static final double sdAllowedMinGap = 0.9;
  static final double sdAllowedMaxGap = 1.1;
  static final double sdDeniedMinGap = 0.8;
  static final double sdDeniedMaxGap = 1.2;
  static final double cdParameter_beta = 0.0001;

  //변수
  private int iLine_size;
  private double dSlope;
  private double sdAvgLineSize;
  private double sdAvgCubeLineSize;
  private Vec4i vPoint;

  //생성자
  public CompareableLine(){}
  public CompareableLine(Vect4i point){}

  //내부 함수
  public Vec4i getPoint()
  public int getLine_size()
  public double getFunctionD(double)
  public double getFunctionS(double)
  public double getSlope()
  public boolean calParams()

  //전역 함수
  public static void setAvgLineSize(double)
  public static void setAvgCubeLineSize(double)
  public static double getAvgLineSize()
}
